### ISSUE TITLE
Windows & Linux: Support `set_dock_menu` equivalents across platforms in a unified way

### ISSUE THREAD

**Issue by versecafe** [zed-industries/zed#12068]
### Description

Add support for `set_dock_menu` to Windows and Linux called `Dock : MacOS`, `Desktop Launcher Actions : Linux`, `Taskbar Customization : Windows`

### Windows
Called *User Tasks* form a menu that supports
- icon
- title
- description
- action (program + arguments)

### Linux ~
Writing to `.desktop` file allows custom actions within *some* linux desktop environments
- title
- action (program + arguments)

More information at [freedesktop.org](https://specifications.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html) # Table 2. Standard Keys

## Key Goal
Avoid fragmenting implementations of this into a per OS custom system and falling into the same pit electron has when handling dock menus

### Linux Sample
![image](https://github.com/zed-industries/zed/assets/147033096/eee763b9-10fc-4186-bd8f-5a054028623e)

### Windows Sample
![image](https://github.com/zed-industries/zed/assets/147033096/0f99d0d2-a552-487f-bf3d-fd034196dab1)

**Comment by SomeoneToIgnore** [comment]
https://github.com/zed-industries/zed/pull/9919 (and the PR it follows-up) 
https://github.com/zed-industries/zed/pull/12067
established the API as

https://github.com/zed-industries/zed/blob/ba1d28f160459bc13c6ec9f044afff1f97fae294/crates/gpui/src/platform.rs#L138-L139


So far, only macOS code provides impls for the methods, but nominally the issue is implemented already — there's a unified way to set menu items and recently opened files.
Other platforms are sure welcome to add those, and that sure happens eventually, so should we close the issue?

**Comment by bbb651** [comment]
Unfortunately, I don't think it's possible to add desktop [actions](https://specifications.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html#extra-actions) dynamically on linux, they are the equivalent of the dock menu but they work more like iOS when you long press an app on the homescreen, in that they are accessible even when Zed is closed. They are currently only used [once](https://github.com/zed-industries/zed/blob/a5b14de4012beebec512a6fc8c30d3638eed7d1b/crates/zed/src/zed.rs#L673) with a constant value, so it should be simple to add it (for reference, this is [vscode's desktop file](https://github.com/zed-industries/zed/blob/a5b14de4012beebec512a6fc8c30d3638eed7d1b/crates/zed/resources/zed.desktop) with the same action), avoiding duplicating action definitions and having it work for arbitrary GPUI applications sounds complicated and depends on their build/packaging scripts...

For recent files, I found this ancient [Recent File Storage Specification](https://www.freedesktop.org/wiki/Specifications/recent-file-spec/) which sounds like it might be used by desktops, but then this [Desktop Bookmark Specification](https://www.freedesktop.org/wiki/Specifications/desktop-bookmark-spec/) seems to replace it. It could've sworn this was supported on gnome but [apparently not](https://gitlab.gnome.org/GNOME/gnome-shell/-/issues/6398), I'm still trying to decipher what spec is used, I found [this gnome extension](https://github.com/maoschanz/quicklists-gnome-shell-extension/) that uses [Gtk.RecentManager](https://docs.gtk.org/gtk4/class.RecentManager.html) which uses gio's `recent://`, and on the KDE side I found [this](https://invent.kde.org/plasma/plasma-desktop/-/blob/master/applets/taskmanager/package/contents/ui/ContextMenu.qml#L108) which and the [backend code it's referring to](https://invent.kde.org/plasma/plasma-desktop/-/blob/master/applets/taskmanager/plugin/backend.cpp#L307).

**Comment by JunkuiZhang** [comment]
Like on Linux, it is also pretty complicated on Windows. On macOS, dock menu acts like a button, on Windows, it is something like:
```rust
Command::new("Zed.exe").with_args("...")
```

It tries to open a new instance of the app, the only infomation is the args we pass.

**Comment by dovakin0007** [comment]
I found out some docs for creating jumplist on windows https://learn.microsoft.com/en-us/samples/microsoft/windows-universal-samples/jumplist/ this is a sample provided by Microsoft to create a sample jump list for an app

**Comment by dovakin0007** [comment]
Is this something similar to the set_dock_menu should I achieve in windows 
![Screenshot 2024-07-28 222636](https://github.com/user-attachments/assets/7a947b8a-5a86-4971-80a3-5636ea8a4b9d)

**Comment by JunkuiZhang** [comment]
> Is this something similar to the set_dock_menu should I achieve in windows

Like I said previously, this will open a new instance of the app. It's not the same thing on macOS which acts like a button.

**Comment by versecafe** [comment]
@JunkuiZhang That could be handled by passing along a flag that tells GPUI to merge the instances a bit hacky but it's a way around windows having poor handling

**Comment by dovakin0007** [comment]
> > Is this something similar to the set_dock_menu should I achieve in windows
> 
> As I mentioned before, this will open a new instance of the app, unlike on macOS where it behaves like a button.
> 
> If you don’t believe me, please give it a try.

https://github.com/zed-industries/zed/pull/15371 seems like you are already working on it

**Comment by JunkuiZhang** [comment]
Yes, this approach is somewhat hacky, but Chrome handles it in a similar way.

**Comment by dovakin0007** [comment]
![image](https://github.com/user-attachments/assets/375fed48-9c4c-4889-8a67-f8160db41ee6)
so is this a part of this feature?

**Comment by JosephTLyons** [comment]
Hello,

We’re closing this issue as part of a cleanup of older feature requests.

Going forward, new feature requests start as GitHub Discussions and only become Issues once the direction is agreed on. This helps ensure that Issues represent scoped, ready-to-build work and makes it easier for contributors to find work that’s ready to pick up. You can read more in our [Feature Request Guidelines](https://github.com/zed-industries/zed/discussions/47963).

As part of this transition, we’re closing feature request Issues that are older than 6 months and have fewer than 10 upvotes.

If this feature hasn’t been implemented and is still important to you, we encourage you to open a new feature request Discussion [here](https://github.com/zed-industries/zed/discussions/new?category=feature-requests). That’s where we track and evaluate new requests.

Thanks for helping shape the future of Zed.
